internal/pkg/template/templates/cicd/buildspec.yml

```diff
@@ -85,49 +85,51 @@ phases:
       - >
         for workload in $WORKLOADS; do
           manifest=$(cat $CODEBUILD_SRC_DIR/copilot/$workload/manifest.yml | ruby -ryaml -rjson -e 'puts JSON.pretty_generate(YAML.load(ARGF))')
-          image_location=$(echo $manifest | jq '.image.location')
-          if [ ! "$image_location" = null ]; then
-            echo "skipping image building because location is provided as $image_location";
-            continue
-          fi
-          base_dockerfile=$(echo $manifest | jq '.image.build')
-          build_dockerfile=$(echo $manifest| jq 'if .image.build?.dockerfile? then .image.build.dockerfile else "" end' | sed 's/"//g')
-          build_context=$(echo $manifest| jq 'if .image.build?.context? then .image.build.context else "" end' | sed 's/"//g')
-          build_target=$(echo $manifest| jq 'if .image.build?.target? then .image.build.target else "" end' | sed 's/"//g')
-          dockerfile_args=$(echo $manifest | jq 'if .image.build?.args? then .image.build.args else "" end | to_entries?')
-          build_cache_from=$(echo $manifest | jq 'if .image.build?.cache_from? then .image.build.cache_from else "" end')
-          df_rel_path=$( echo $base_dockerfile | sed 's/"//g')
-          if [ -n "$build_dockerfile" ]; then 
-            df_rel_path=$build_dockerfile
-          fi
-          df_path=$df_rel_path
-          df_dir_path=$(dirname "$df_path")
-          if [ -n "$build_context" ]; then
-            df_dir_path=$build_context
-          fi
-          build_args=
-          if [ -n "$dockerfile_args" ]; then
-            for arg in $(echo $dockerfile_args | jq -r '.[] | "\(.key)=\(.value)"'); do 
-              build_args="$build_args--build-arg $arg "
-            done
-          fi
-          if [ -n "$build_target" ]; then
-            build_args="$build_args--target $build_target "
-          fi
-          if [ -n "$build_cache_from" ]; then
-            for arg in $(echo $build_cache_from | jq -r '.[]'); do
-              build_args="$build_args--cache-from $arg "
-            done
-          fi
-          echo "Name: $workload"
-          echo "Relative Dockerfile path: $df_rel_path"
-          echo "Docker build context: $df_dir_path"
-          echo "Docker build args: $build_args"
-          echo "Running command: docker build -t $workload:$tag $build_args-f $df_path $df_dir_path";
-          docker build -t $workload:$tag $build_args-f $df_path $df_dir_path;
-          image_id=$(docker images -q $workload:$tag);
           for env in $pl_envs; do
-            repo=$(cat $CODEBUILD_SRC_DIR/infrastructure/$workload-$env.params.json | jq '.Parameters.ContainerImage' | sed 's/"//g');
+            tag=$(sed 's/:/-/g' <<<"${CODEBUILD_BUILD_ID}-${env}" | rev | cut -c 1-128 | rev)
+            image=$(echo $manifest | jq ".image * (.environments?.$env?.image? + {})")
```

Comment:
This looks cool! Though I'm not very sure if it has the same behavior as our environment override (we are not doing a very good job in this buildspec as well 🙇 ). Could you help me to confirm if it works for the following scenarios?

- we use every field that is set in  `env.image` to update or insert the corresponding field in `image`
-  if `image.build` and `env.image.location` is set at the same time, we'll use `env.image.location`. Vise versa

Please let me know if anything doesn't make sense to you 🙏

Reply:
This is awesome. 

The usage of `*` is super neat! However we might need to include an extra logic to address @iamhopaul123 's second concern. To add on top of iamhopaul123's comment -  consider this manifest:
```yaml
image:
   location: <an URL>

environments:
   test:
      image:
         // We want to build an image instead of using an existing one for the test environment
         build: 
             dockerfile: ./Dockerfile
```
After the `*` operator, which should be performing an upsert (right?), the output of `image=$(echo $manifest | jq ".image * (.environments?.$env?.image? + {})")` would be a struct that's equivalent to
```yaml
image:
   build: 
        dockerfile: ./Dockerfile
   image:
         location: <an URL>      
```
Next thing we do will be skip building the image because `image.location` is not null; however, the expected behavior should be using `image.build` to build and push a new image.

Therefore, we need to include an extra logic here, that unset `image.build` if `image.location` is specified in the environment override section, and vice versa.

Reply:
@iamhopaul123 Thanks for reply.

> we use every field that is set in `env.image` to update or insert the corresponding field in `image`

Right. Here is a sample:
```
$ echo '{"sample": {"text": "bar", "aaa": "bbb"}, "another": {"sample": {"aaa": "BBB", "ccc": "ddd"}}}' | jq '.sample * .another.sample'
{
  "text": "bar",
  "aaa": "BBB",
  "ccc": "ddd"
}
```

> if `image.build` and `env.image.location` is set at the same time, we'll use `env.image.location`. Vise versa

Yes. In this case, docker image build will be skipped.

@Lou1415926 It will be merged as following, I think.
```
image:
   build: 
        dockerfile: ./Dockerfile
   location: <an URL>
```

> Therefore, we need to include an extra logic here, that unset `image.build` if `image.location` is specified in the environment override section, and vice versa.

I see. Are there some documentation or implementation about this logic?

Reply:
I've thought about all the patterns, is this the behavior you are expecting?

- `image.location` * `env.image.location` -> Just merge (overwrite)
- `image.location` * `env.image.build` -> Unset `image.location` and use `image.build`
- `image.build` * `env.image.location` -> Just merge (`image.build` will be ignored)
- `image.build` * `env.image.build` -> Just merge ?

Or should I take other parameters (`credentials`, `labels`, `depends_on`) into consideration?

Reply:
Thanks for following up! Yes it is documented [here](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#image-location) that says _"`image.location`... Mutually exclusive with image.build"_. We should add the same note for `image.build`'s doc as well - our bad! This logic is implemented in the internal code but unfortunately we can't easily reuse that code for buildspec's image building  : (

The other parameters are fine - only `image.location` * `image.build` needs the special treatment. Your summary sounds good to me!

Reply:
Thanks for reply!
I updated merge logic to ignore `.image.location` if `.environments.<env>.image` exists.